Add GitHub action workflow to purge webhooks from this repo weekly

nfagerlund opened the pull request:
This action is copied from the hashicorp/terraform-random-1 repo, and was written by @brandonc. These repos are used for the same purpose but by different tests, and are vulnerable to the same orphaned webhook problems in cases where tests fail catastrophically enough to prevent cleanup from running.

Since this repo doesn't appear to have any valid persistent webhooks (only temporary ones created by tests running in CI), deleting all hooks weekly seems appropriate.

Co-authored-by: @brandonc

nfagerlund committed:
Add GitHub action workflow to purge webhooks from this repo weekly

This action is copied from the hashicorp/terraform-random-1 repo, and was
written by @brandonc. These repos are used for the same purpose but by different
tests, and are vulnerable to the same orphaned webhook problems in cases where
tests fail catastrophically enough to prevent cleanup from running.

Since this repo doesn't appear to have any valid persistent webhooks (only
temporary ones created by tests running in CI), deleting all hooks weekly seems
appropriate.

Co-authored-by: @brandonc

nfagerlund commented:
For reviewers: Make sure I changed the right api paths and the action isn't trying to access a different repo like terraform-random-1. :)

brandonc commented on .github/workflows/cron_weekend.yaml:
```diff
@@ -0,0 +1,38 @@
+  schedule:
+    - cron: '0 0 * * SAT'
+
+jobs:
+  fetchWebhooks:
+    runs-on: ubuntu-latest
+    outputs:
+      hook_ids: ${{ steps.format.outputs.hook_ids }}
+
+    steps:
+      - uses: octokit/request-action@v2.x
+        id: request
+        with:
+          route: GET /repos/hashicorp/test-policy-set/hooks
+        env:
+          GITHUB_TOKEN: ${{ secrets.GH_TOKEN }}
+      - id: format
+        run: |
+          echo "hook_ids=$(jq -n '${{ steps.request.outputs.data }}' | jq -rc '[.[] .id]')\n" >> $GITHUB_OUTPUT
+      - run: |
+          echo 'Deleting these webhook ids: ${{ steps.format.outputs.hook_ids }}'
+  cleanWebhooks:
+    runs-on: ubuntu-latest
+    needs: [ fetchWebhooks ]
+    if: fromJSON(needs.fetchWebhooks.outputs.hook_ids)[0] != null
+    strategy:
+      matrix:
+        value: ${{fromJSON(needs.fetchWebhooks.outputs.hook_ids)}}
+    steps:
+      - uses: octokit/request-action@v2.x
+        id: delete_webhook
+        with:
+          route: DELETE /repos/hashicorp/test-policy-set/hooks/${{ matrix.value }}
+        env:
+          GITHUB_TOKEN: ${{ secrets.GH_TOKEN }}
```
I modified the source repo to allow it to be called from any other repo. You could do it this way, too (I think)

Here's the docs https://docs.github.com/en/actions/using-workflows/reusing-workflows#calling-a-reusable-workflow

```suggestion
jobs:
  call-delete-all-webhooks:
    steps:
      - uses: hashicorp/terraform-random-1/.github/workflows/delete_all_webhooks.yml@main
        secrets:
          GH_TOKEN: ${{ secrets.GH_TOKEN }}
```

nfagerlund replied:
Ah yeah, I quite prefer that! I was thinking we would probably publish an action once we were happy with the way it works, but this is a nice lightweight way to dedupe for now.

nfagerlund committed:
Re-use the workflow from hashicorp/terraform-random-1

Viable as of https://github.com/hashicorp/terraform-random-1/commit/cc792272c34b6fcc561a57465b93291c738dfc90

Co-authored-by: Brandon Croft <brandon.croft@gmail.com>

nfagerlund commented:
We also just added the GH_TOKEN secret required for the action to work.
